### PR TITLE
Add {valid, full} option for Pooling layer

### PR DESCRIPTION
**Pull request by Darwin2011**
Mxnet's pooling layer's output doesn't match Caffe's pooling's output, although Mxnet's convolution layer's output shape is same as Caffe. This issue will cause Googlenet from caffe convertor failing.

The following links shows pooling output size formula.
for caffe, 
https://github.com/BVLC/caffe/blob/master/src/caffe/layers/pooling_layer.cpp#L90
for mxnet, 
https://github.com/dmlc/mxnet/blob/master/src/operator/pooling-inl.h#L202

And please see #2999. @winstywang @ChenFengAndy

So I add two options for pooling layer {valid, same}. valid is for current mxnet's pooling output and for caffe, it uses same option
1) default pooling option is valid, same as current mxnet.
2) if model is from caffe, same option for pooling layer will be on.

Please give some comments. Thanks a lot.

### CODE REVIEW

**Comment by sxjscience** [comment]
Why call it `same`?

**Comment by Darwin2011** [comment]
@sxjscience 
Maybe "full" is one better name? I will change it to "full". That makes more sense.

**Comment by sxjscience** [comment]
@Darwin2011 Could we manually adjust the padding to achieve the similar result?

**Comment by Darwin2011** [comment]
@sxjscience Could you please explain a little bit more? Thanks.

**Comment by sxjscience** [comment]
@Darwin2011 I'm not sure whether this will break the C++ implementations.

**Comment by sxjscience** [comment]
@antinucleon @piiswrong @winstywang What do you think about this? Some old caffemodels like VGG_CNN_M requires this type of compatible pooling. Another solution is to fine-tune these networks using the revised structure (pad 1 in the pooling layer) and add them to our own model zoo.

**Comment by Darwin2011** [comment]
@sxjscience 
Caffe model converter also cannot translate bvlc googlenet on my side.

**Comment by sxjscience** [comment]
@Darwin2011 Yes, this is a problem. One way to deal with the problem is to add `pad=(1,1)` for lower level pooling layers like here (https://github.com/dmlc/mxnet/blob/master/example/image-classification/symbol_inception-bn-28-small.py#L19-L20). However, we may need to fine-tune the network again since the pooling function has been changed.
BTW, have you tested if the pooling layer works well for the `SAME` mode?

**Comment by Darwin2011** [comment]
@sxjscience 
I haven't tested it. What I only test is in "same" mode is on, Caffe convertor model can work well and mxnet can get same accuracy as Caffe. Let me test it with cifar10 training.

Thanks.

**Comment by piiswrong** [comment]
I think this should be added. Although the name "same" sounds cryptic. How about "compatible"?

File: src/operator/pooling-inl.h [hunk 1]
```diff
@@ -25,13 +25,15 @@ namespace pool_enum {
 enum PoolingOpInputs {kData};
 enum PoolingOpOutputs {kOut};
 enum PoolingOpType {kMaxPooling, kAvgPooling, kSumPooling};
+enum PoolingOpPadConventionType {kSame, kValid};
```
Comment (piiswrong) [inline src/operator/pooling-inl.h]: put kValid first

File: src/operator/pooling-inl.h [hunk 2]
```diff
@@ -47,6 +49,11 @@ struct PoolingParam : public dmlc::Parameter<PoolingParam> {
     .add_enum("avg", pool_enum::kAvgPooling)
     .add_enum("sum", pool_enum::kSumPooling)
     .describe("Pooling type to be applied.");
+ 
+    DMLC_DECLARE_FIELD(pad_convention).set_default(pool_enum::kValid)
+    .add_enum("same", pool_enum::kSame)
```
Comment (piiswrong) [inline src/operator/pooling-inl.h]: change same to "full" or "compatible" and add document describing the difference (round up or down, discard extra or pad zero) and say that use full for compatibility with caffe

**Comment by sbodenstein** [comment]
@piiswrong @Darwin2011 @sxjscience: These names are really bad. In the literature, 'same', 'valid' and 'full' have precise meanings for pooling and convolutions in the literature:

- same: the tensor is padded so that the input and output tensors are the same width and height. 
- valid: no padding

This is used by [TensorFlow](https://www.tensorflow.org/versions/r0.10/api_docs/python/nn.html#convolution) or [Numpy](http://docs.scipy.org/doc/numpy/reference/generated/numpy.convolve.html), for example. Using these names for this will cause confusion (esp. if MXnet decided to add these very useful options for convolution + pooling).

How about "caffe_compatibibility" which is a bool? 

Here is a question that we should answer first: what is the actual computational difference in the pooling layer? Is the MXNet pooling layer equivalent to Caffe with some extra padding for certain input dimensions? If so, we can fix this at the Caffe importer level rather than hacking it into the MXNet pooling definition.

**Comment by piiswrong** [comment]
the difference is basically round down vs up in division. So valid makes
sense. We just need to name the other one better.

**Comment by Darwin2011** [comment]
@sbodenstein Thanks for your advice.
Please see https://github.com/soumith/convnet-benchmarks/issues/82.
Googlenet symbols fails because of different pooling output size conventions. That's why I add one more options to pooling layer.

**Comment by Darwin2011** [comment]
Thanks a lot for your review, @piiswrong 
renaming done and add some descriptions.

**Comment by sxjscience** [comment]
@Darwin2011 I think we can merge this in if the test passes.

**Comment by piiswrong** [comment]
please update
